core/server/session_logs.go

```diff
@@ -114,35 +126,191 @@ func (cs *coreServer) GetSessionLogs(ctx context.Context, msg *pb.GetSessionLogs
+	sort.Slice(logEntries, func(i, j int) bool {
+		tsi, _ := time.Parse(time.RFC3339, logEntries[i].GetTimestamp())
+		tsj, _ := time.Parse(time.RFC3339, logEntries[j].GetTimestamp())
+		return tsi.Before(tsj)
+	})
 
-	for _, log := range logs {
-		logEntry := &pb.LogEntry{}
-		err = json.Unmarshal([]byte(log), &logEntry)
+	return &pb.GetSessionLogsResponse{
+		Logs:       logEntries,
+		NextToken:  firstToken + "," + secondToken,
+		LogSources: append([]string{logger.SessionLogSource}, logSources...),
+	}, nil
+}
+
+func detectLogLevel(message string) string {
+	message = strings.ToLower(message)
+	errorRegex := regexp.MustCompile(`(err(or)?|fatal|ftl)`)
+	warnRegex := regexp.MustCompile(`(warn(ing)?|wrn)`)
+
+	if errorRegex.MatchString(message) {
+		return "error"
+	} else if warnRegex.MatchString(message) {
+		return "warn"
+	} else {
+		return "info"
+	}
+}
+
+type PodLog struct {
+	Date       time.Time `json:"date"`
+	Time       time.Time `json:"time"`
+	Log        string    `json:"log"`
+	Msg        string    `json:"msg"`
+	Kubernetes struct {
+		PodName        string            `json:"pod_name"`
+		NamespaceName  string            `json:"namespace_name"`
+		PodID          string            `json:"pod_id"`
+		Labels         map[string]string `json:"labels"`
+		Annotations    map[string]string `json:"annotations"`
+		Host           string            `json:"host"`
+		ContainerName  string            `json:"container_name"`
+		DockerID       string            `json:"docker_id"`
+		ContainerHash  string            `json:"container_hash"`
+		ContainerImage string            `json:"container_image"`
+	} `json:"kubernetes"`
+}
+
+func getPodLogs(ctx context.Context, nextToken string, minioClient s3Reader, bucketName string, logSourceFilter string, logLevelFilter string) ([]*pb.LogEntry, string, []string, error) {
+	// we use the second part of the token as the startAfter value
+	if strings.Contains(nextToken, ",") {
+		parts := strings.SplitN(nextToken, ",", 2)
+		nextToken = parts[1]
+	}
+
+	var (
+		logs      []*pb.LogEntry
+		lastToken string
+	)
+
+	tmpLogSources := map[string]struct{}{}
+	for obj := range minioClient.ListObjects(ctx, bucketName, minio.ListObjectsOptions{
+		Prefix:     "fluent-bit-logs",
+		StartAfter: nextToken,
+		Recursive:  true,
+	}) {
+		if obj.Err != nil {
+			return nil, "", nil, obj.Err
+		}
+
+		content, err := minioClient.GetObject(ctx, bucketName, obj.Key, minio.GetObjectOptions{})
+		if err != nil {
+			return nil, "", nil, err
+		}
+
+		b, err := io.ReadAll(content)
 		if err != nil {
-			return nil, err
+			return nil, "", nil, err
+		}
+
+		messages := strings.Split(string(b), "\n")
+		for _, message := range messages {
+			if strings.TrimSpace(message) == "" {
+				continue
+			}
+
+			podLog := &PodLog{}
+			if err := json.Unmarshal([]byte(message), podLog); err != nil {
+				return nil, "", nil, err
+			}
+
+			innerMessage := podLog.Log
+			if innerMessage == "" {
+				innerMessage = podLog.Msg
```

Comment:
❔ What does `podLog.Msg` contain? Is it some sort of a system/error message?

Reply:
Different log formats have different ways of storing their inner messages. Some programs store inner messages in "log" field, some store them in the "msg" field. 

> How does the inner message in the logs work?
We do not and will not know before hand what field like this would appear in the future. 

If something is broken, we have to add support more field names here.

Reply:
Got it, thanks!